File: Algorithms/src/2975. Maximum Square Area by Removing Fences From a Field/Code/solution.cpp
class Solution
{
public:
    int maximizeSquareArea(int m, int n, vector<int> &hFences, vector<int> &vFences)
    {
        const long long MOD = 1e9 + 7;

        // Add boundary fences
        hFences.push_back(1);
        hFences.push_back(m);
        vFences.push_back(1);
        vFences.push_back(n);

        sort(hFences.begin(), hFences.end());
        sort(vFences.begin(), vFences.end());

        unordered_set<long long> horizontal;
        unordered_set<long long> vertical;

        // All possible horizontal distances
        for (int i = 0; i < hFences.size(); i++)
        {
            for (int j = i + 1; j < hFences.size(); j++)
            {
                horizontal.insert(hFences[j] - hFences[i]);
            }
        }

        // All possible vertical distances
        for (int i = 0; i < vFences.size(); i++)
        {
            for (int j = i + 1; j < vFences.size(); j++)
            {
                vertical.insert(vFences[j] - vFences[i]);
            }
        }

        long long maxSide = 0;

        // Find largest common distance
        for (auto d : horizontal)
        {
            if (vertical.count(d))
            {
                maxSide = max(maxSide, d);
            }
        }

        if (maxSide == 0)
            return -1;

        return (maxSide * maxSide) % MOD;
    }
};

File: Algorithms/src/2975. Maximum Square Area by Removing Fences From a Field/Code/solution.py
class Solution:
    def maximizeSquareArea(self, m: int, n: int, hFences: list[int], vFences: list[int]) -> int:
        MOD = 10**9 + 7

        hFences.append(1)
        hFences.append(m)
        vFences.append(1)
        vFences.append(n)

        hFences.sort()
        vFences.sort()

        horizontal = set()
        vertical = set()

        for i in range(len(hFences)):
            for j in range(i + 1, len(hFences)):
                horizontal.add(hFences[j] - hFences[i])

        for i in range(len(vFences)):
            for j in range(i + 1, len(vFences)):
                vertical.add(vFences[j] - vFences[i])

        maxSide = 0
        for d in horizontal:
            if d in vertical:
                maxSide = max(maxSide, d)

        if maxSide == 0:
            return -1

        return (maxSide * maxSide) % MOD
